### PR TITLE
Consul dataplane metrics 

### PR DESCRIPTION
**Pull request by chapmanc**
This PR simply adds a consul dataplane metrics server. A subsequent PR will be added to add in the specified metrics requested by the RFC.

This PR is best viewed commit by commit: 

1. 9783fca4a6697558d2e4654c6f1c6412fba855e4 adds in the necesseary go-metrics and prometheus modules 
2. 98cc9efe75b88de7557a983a6bb33f784fa0ddfa makes the envoy admin url which we scrape for merged metrics use the configurable values that we allow in consul dataplane vs a hard coded value 
3. 296dbf58b46f65320a93bb7848e58da3e7155aa9 refactors the lifecycle controls and simplifies the merged metrics handler to take in a list of urls to scrape
4. 8af9af87f48a1cc0531ad4e03401a96eadedfd81 adds in a metrics server for consul dataplane that will allow the metrics to be scraped by the merged metrics handler
5. 49bf1dbfbacca3fb120036b045af78a66239723c fixes the tests

### CODE REVIEW

File: pkg/consuldp/metrics.go [hunk 1]
```diff
@@ -168,3 +219,74 @@ func (m *metricsConfig) scrapeError(rw http.ResponseWriter, url string, err erro
...
+func (m *metricsConfig) getPromDefaults() (*prom.Registry, *prometheus.PrometheusOpts, error) {
+	r := prom.NewRegistry()
+	reg := prom.WrapRegistererWithPrefix("consul_dataplane_", r)
+	err := reg.Register(collectors.NewGoCollector())
+	if err != nil {
+		return nil, nil, err
+	}
+	opts := &prometheus.PrometheusOpts{
+		Registerer: reg,
+		// GaugeDefinitions: ,
```
Comment (chapmanc) [inline pkg/consuldp/metrics.go:236]: This where we will register the metrics from other packages etc.

File: pkg/consuldp/metrics.go [hunk 2]
```diff
@@ -8,106 +8,165 @@ import (
...
-	httpServer  *http.Server
-	client      httpGetter
+	logger hclog.Logger
+
+	cfg                *TelemetryConfig
+	envoyAdminAddr     string
+	envoyAdminBindPort int
+
+	// merged metrics config
+	promScrapeServer *http.Server // the server that will will serve all the merged metrics
```
Comment (riddhi89) [inline pkg/consuldp/metrics.go:52]:
```suggestion
	promScrapeServer *http.Server // the server that will serve all the merged metrics
```

**Review by riddhi89 — APPROVED** [review]
Great job! LGTM

**Comment by boxofrad** [comment]
Just wanted to say thanks for breaking this up into digestible commits! It made reviewing this a breeze 💨

File: pkg/consuldp/metrics_test.go [hunk 3]
```diff
@@ -93,15 +121,15 @@ func TestMetricsServerEnabled(t *testing.T) {
 
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
-			m.startMetrics(ctx, &bootstrap.BootstrapConfig{PrometheusBindAddr: "nonempty"})
-			require.Equal(t, metricsBackendBindAddr, m.httpServer.Addr)
+			_ = m.startMetrics(ctx, &bootstrap.BootstrapConfig{PrometheusBindAddr: "nonempty"})
```
Comment (pglass) [inline pkg/consuldp/metrics_test.go:124]: nit: we could check this error (`require.NoError`)

**Review by pglass — APPROVED** [review]
Looks good

**Pull request merged**